Zend/zend_attributes.c

```diff
@@ -91,6 +100,91 @@ ZEND_METHOD(AllowDynamicProperties, __construct)
 	ZEND_PARSE_PARAMETERS_NONE();
 }
 
+ZEND_METHOD(SensitiveParameter, __construct)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+}
+
+ZEND_METHOD(SensitiveParameterValue, __construct)
+{
+	zval *value;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ZVAL(value)
+	ZEND_PARSE_PARAMETERS_END();
+
+	ZVAL_COPY(OBJ_PROP_NUM(Z_OBJ_P(ZEND_THIS), 0), value);
+}
+
+/* {{{ */
+ZEND_METHOD(SensitiveParameterValue, getValue)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	ZVAL_COPY(return_value, OBJ_PROP_NUM(Z_OBJ_P(ZEND_THIS), 0));
+} /* }}} */
+
+/* {{{ */
+ZEND_METHOD(SensitiveParameterValue, __debugInfo)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	RETURN_EMPTY_ARRAY();
+} /* }}} */
+
+/* {{{ */
+ZEND_METHOD(SensitiveParameterValue, __serialize)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	RETURN_EMPTY_ARRAY();
+} /* }}} */
+
+/* {{{ */
+ZEND_METHOD(SensitiveParameterValue, __unserialize)
+{
+	HashTable *data;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "h", &data) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	zend_throw_exception(NULL, "Unserializing objects of class SensitiveParameterValue is not supported.", 0);
+	RETURN_THROWS();
+} /* }}} */
+
+static zend_object *attributes_sensitive_parameter_value_new(zend_class_entry *ce)
+{
+	attributes_sensitive_parameter_value *intern = zend_object_alloc(sizeof(attributes_sensitive_parameter_value), ce);
+	memset(intern, 0, sizeof(attributes_sensitive_parameter_value) - sizeof(zend_object));
+
+	zend_object_std_init(&intern->std, ce);
+	object_properties_init(&intern->std, ce);
+	intern->std.handlers = &attributes_object_handlers_sensitive_parameter_value;
+
+	return &intern->std;
+}
+
+static inline attributes_sensitive_parameter_value *attributes_sensitive_parameter_value_from_obj(zend_object *obj) {
+	return (attributes_sensitive_parameter_value *)((char *)(obj) - XtOffsetOf(attributes_sensitive_parameter_value, std));
+}
+
+static void attributes_sensitive_parameter_value_free_obj(zend_object *object)
+{
+	attributes_sensitive_parameter_value *v = attributes_sensitive_parameter_value_from_obj(object);
+
+	zend_object_std_dtor(&v->std);
+}
+
+static HashTable *attributes_sensitive_parameter_value_get_properties_for(zend_object *zobj, zend_prop_purpose purpose)
+{
+	if (purpose == ZEND_PROP_PURPOSE_ARRAY_CAST) {
+		return NULL;
+	} else {
+		return zend_std_get_properties_for(zobj, purpose);
```

Comment:
var_export/debug_zval_dump use ZEND_PROP_PURPOSE_DEBUG, var_export output may be logged

You may want to always return null, or zend_empty_array, they both behave about the same way

```c
//ext/ffi/ffi.c
static HashTable *zend_fake_get_properties(zend_object *obj) /* {{{ */
{
	return (HashTable*)&zend_empty_array;
}
``` 

```c
typedef enum _zend_prop_purpose {
	/* Used for debugging. Supersedes get_debug_info handler. */
	ZEND_PROP_PURPOSE_DEBUG,
	/* Used for (array) casts. */
	ZEND_PROP_PURPOSE_ARRAY_CAST,
	/* Used for serialization using the "O" scheme.
	 * Unserialization will use __wakeup(). */
	ZEND_PROP_PURPOSE_SERIALIZE,
	/* Used for var_export().
	 * The data will be passed to __set_state() when evaluated. */
	ZEND_PROP_PURPOSE_VAR_EXPORT,
	/* Used for json_encode(). */
	ZEND_PROP_PURPOSE_JSON,
	/* Dummy member to ensure that "default" is specified. */
	_ZEND_PROP_PURPOSE_NON_EXHAUSTIVE_ENUM
} zend_prop_purpose;
```

Reply:
Thanks, I've added var_export() and debug_zval_dump() to the test: b51d8241f4c3eecd0cb3280969b6852b0b6f792c. The latter already was protected (I assume due to `__debugInfo()`).

FWIW: `return (HashTable*)&zend_empty_array;` didn't work. That causes a segmentation fault. I've used `NULL`.

Reply:
> FWIW: return (HashTable*)&zend_empty_array; didn't work. That causes a segmentation fault. I've used NULL.

Oh, I see why. ext/ffi/ffi.c overrides both get_properties and get_debug_info so zend_std_get_properties_for uses get_debug_info instead, which by design always returns a refcounted (non-immutable) array.

Returning null in `get_properties_for` seems to work for all use cases I checked.

(GC_PROTECT_RECURSION unconditionally modifies flags, GC_TRY_PROTECT_RECURSION)


```c
// zend_ffi_cdata_get_debug_info
		case ZEND_FFI_TYPE_FUNC:
			ht = zend_new_array(0);
			// TODO: function name ???
			*is_temp = 1;
			return ht;
```

```c
	case IS_OBJECT:
		myht = zend_get_properties_for(struc, ZEND_PROP_PURPOSE_DEBUG);
		if (myht) {
			if (GC_IS_RECURSIVE(myht)) {
				PUTS("*RECURSION*\n");
				zend_release_properties(myht);
				return;
			}
			GC_PROTECT_RECURSION(myht);
		}
```

```
==103531==                                                                                                                     
==103531== Process terminating with default action of signal 11 (SIGSEGV)                                                      
==103531==  Bad permissions for mapped region at address 0x1106CA4
==103531==    at 0x66595A: php_debug_zval_dump (var.c:353)                                                                     
==103531==    by 0x665FA9: zif_debug_zval_dump (var.c:413)                                                                     
==103531==    by 0x7C2004: ZEND_DO_ICALL_SPEC_RETVAL_UNUSED_HANDLER (zend_vm_execute.h:1250)                                   
==103531==    by 0x83FA69: execute_ex (zend_vm_execute.h:55499)                                                                
==103531==    by 0x845313: zend_execute (zend_vm_execute.h:59858)
==103531==    by 0x780073: zend_execute_scripts (zend.c:1744)  
==103531==    by 0x6C2014: php_execute_script (main.c:2535)                                                                    
==103531==    by 0x907155: do_cli (php_cli.c:965) 
==103531==    by 0x908435: main (php_cli.c:1367) 
```